// ton-contracts/dist/contracts/messages/HighloadWalletMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HighloadWalletMessage = void 0;
const ton_1 = require("ton");
class HighloadWalletMessage {
    seqno;
    walletId;
    timeout;
    messages;
    constructor(args) {
        this.seqno = args.seqno;
        this.walletId = args.walletId;
        this.messages = args.messages;
        this.timeout = args.timeout;
    }
    writeTo(cell) {
        // Message
        cell.bits.writeUint(this.walletId, 32);
        cell.bits.writeUint(this.timeout, 32);
        cell.bits.writeUint(this.seqno, 32);
        // Dict
        if (this.messages.length > 0) {
            const dict = new Map();
            for (let i = 0; i < this.messages.length; i++) {
                dict.set('' + i, this.messages[i]);
            }
            const dictCell = (0, ton_1.serializeDict)(dict, 16, (src, cell) => {
                cell.bits.writeUint8(src.sendMode);
                const d = new ton_1.Cell();
                src.order.writeTo(d);
                cell.refs.push(d);
            });
            cell.bits.writeBit(true);
            cell.refs.push(dictCell);
        }
        else {
            cell.bits.writeBit(false);
        }
    }
}
exports.HighloadWalletMessage = HighloadWalletMessage;
