// ton-contracts/dist/contracts/WhitelistedWalletSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhitelistedWalletSource = void 0;
const ton_1 = require("ton");
class WhitelistedWalletSource {
    static SOURCE = Buffer.from('te6ccgECDAEAARMAART/APSkE/S88sgLAQIBIAIDAgFIBAUC+PLtRNDTH9P/0//6QCH6RALTH/gjAdEIgwjXGCDTH1EauvKhAfkBVHAo+RBROPkQUSKx8qIBs46rUoKhgQcIufJmUwDXSvKlkyDXSo6R0wcBgQD8sPJk1FQxNNs88qPobCLRBZM4bCHi+AAEpMjLHxPL/8v/Ac8Wyx/J7VQKCwAE0DACASAGBwIBIAgJAE2+WX9qJoaY+Y6f+Y6f+Y/SAY6Y/o/BGA0JBAg4RcysCDhADQ8Bg4QAF7s5ztRNDTHzHXC/+AARuMl+1E0NcLH4ADIC0NMDAXiwk18DcOD6QDH6QDD6RAK6ArqwAB74D5Mg10qW0wfUAvsA6NE=', 'base64');
    static create(opts) {
        // Resolve parameters
        let masterKey = opts.masterKey;
        let restrictedKey = opts.restrictedKey;
        let workchain = opts.workchain;
        let whitelistedAddress = opts.whitelistedAddress;
        // Build initial code and data
        let initialCode = ton_1.Cell.fromBoc(WhitelistedWalletSource.SOURCE)[0];
        let initialData = new ton_1.Cell();
        initialData.bits.writeUint(0, 32); // SeqNo
        initialData.bits.writeBuffer(restrictedKey); // Restricted key
        initialData.bits.writeBuffer(masterKey); // Master key
        initialData.bits.writeAddress(whitelistedAddress); // Whitelisted address
        initialData.bits.writeUint(0, 32); // Last restricted tx
        return new WhitelistedWalletSource({ masterKey, restrictedKey, whitelistedAddress, initialCode, initialData, workchain });
    }
    static restore(backup) {
        const store = new ton_1.ConfigStore(backup);
        return WhitelistedWalletSource.create({
            workchain: store.getInt('wc'),
            restrictedKey: store.getBuffer('pk'),
            masterKey: store.getBuffer('mk'),
            whitelistedAddress: store.getAddress('wa')
        });
    }
    masterKey;
    restrictedKey;
    whitelistedAddress;
    initialCode;
    initialData;
    workchain;
    type = 'org.ton.wallets.whitelisted';
    walletVersion = 'v1'; // Conformance to wallet interface
    constructor(opts) {
        this.masterKey = opts.masterKey;
        this.restrictedKey = opts.restrictedKey;
        this.whitelistedAddress = opts.whitelistedAddress;
        this.initialCode = opts.initialCode;
        this.initialData = opts.initialData;
        this.workchain = opts.workchain;
        Object.freeze(this);
    }
    backup = () => {
        const config = new ton_1.ConfigStore();
        config.setInt('wc', this.workchain);
        config.setBuffer('pk', this.restrictedKey);
        config.setBuffer('mk', this.masterKey);
        config.setAddress('wa', this.whitelistedAddress);
        return config.save();
    };
    describe = () => {
        return 'Whitelisted Restricted Wallet. Address: ' + this.whitelistedAddress.toFriendly() + ', Master Key: ' + this.masterKey.toString('hex') + ', Default Key: ' + this.restrictedKey.toString('hex');
    };
}
exports.WhitelistedWalletSource = WhitelistedWalletSource;
